angr/sim_manager.py

```diff
@@ -0,0 +1,927 @@
+        """Step a stash of states forward and categorize the successors appropriately.
+
+        The parameters to this function allow you to control everything about the
+        stepping and categorization process.
+
+        :param stash:           The name of the stash to step (default: 'active')
+        :param n:               The number of times to step (default: 1 if "until" is not provided)
+        :param selector_func:   If provided, should be a function that takes a state and returns a
+                                boolean. If True, the state will be stepped. Otherwise, it will be
+                                kept as-is.
+        :param step_func:       If provided, should be a function that takes a SimulationManager and
+                                returns a SimulationManager. Will be called with the SimulationManager
+                                at every step. Note that this function should not actually perform any
+                                stepping - it is meant to be a maintenance function called after each step.
+        :param successor_func:  If provided, should be a function that takes a state and return its successors.
+                                Otherwise, project.factory.successors will be used.
+        :param until:           If provided, should be a function that takes a SimulationManager and
+                                returns True or False. Stepping will terminate when it is True.
+
+        Additionally, you can pass in any of the following keyword args for project.factory.sim_run:
+
+        :param jumpkind:        The jumpkind of the previous exit
+        :param addr:            An address to execute at instead of the state's ip.
+        :param stmt_whitelist:  A list of stmt indexes to which to confine execution.
+        :param last_stmt:       A statement index at which to stop execution.
+        :param thumb:           Whether the block should be lifted in ARM's THUMB mode.
+        :param backup_state:    A state to read bytes from instead of using project memory.
+        :param opt_level:       The VEX optimization level to use.
+        :param insn_bytes:      A string of bytes to use for the block instead of the project.
+        :param size:            The maximum size of the block, in bytes.
+        :param num_inst:        The maximum number of instructions.
+        :param traceflags:      traceflags to be passed to VEX. Default: 0
+
+        :returns:           The resulting SimulationManager.
+        :rtype:             SimulationManager
+        """
+        stash = stash or 'active'
+        # 8<----------------- Compatibility layer -----------------
+        if any(_ is not None for _ in (n, selector_func, step_func, successor_func, until)):
+            return self._step_compat(n, selector_func, step_func, stash, successor_func, until, **run_args)
+        # ------------------ Compatibility layer ---------------->8
```

Comment:
See above. while these things are functionality that's duplicated by exploration techniques, they're too useful to relegate to a compatibility layer.

Reply:
This is exactly the `adhoc_technique` method is here for (though, it could use a better name). To clarify this decision, I should mention that the essence of this design is to make the SimulationManager "do one thing" (that is, to manage state stashes, as well, as to step them forward), and "do it well". From this perspective, the `step` method should 

> Step a stash of states forward and categorize the successors appropriately.

and nothing more. Every other influence on the stepping process should come from the exploration techniques. In any case, I would strongly suggest to at least remove the `n` and the `until` arguments from the prototype, as the `step` method is meant to perform a _single_ step in a given stash. To perform many steps, one should use the `run` method.